src/wms/src/main/java/org/geoserver/wms/decoration/LegendDecoration.java

```diff
@@ -185,8 +122,8 @@ public Dimension findOptimalSize(Graphics2D g2d, WMSMapContent mapContext) {
 
     public void paint(Graphics2D g2d, Rectangle paintArea, WMSMapContent mapContext)
             throws Exception {
-        List<LayerLegend> legends = this.legends.get();
-
+        List<LayerLegend> legends =
+                this.legends.get() != null ? this.legends.get() : getLayerLegend(g2d, mapContext);
```

Comment:
Ok, so you need the legends to be computed, but they are normally done in the method above... and are making sure they are present even if the auto-sizing did not happen. Can you add a code comment to clarify it?